### PR TITLE
Config fundamentals

### LINKED ISSUE #763: Document how to use new configuration system
For people coming from .NET Framework, there is a gap of how to do configuration when they are used to app.config. There should be documentation and an example of how to replace it.

### PR DESCRIPTION
**Pull request by IEvangelist**
## Summary

- Port over and adopt "Configuration in .NET" article from the ASP.NET Core docs repo
- Introduce new code projects representing various configuration examples
- Update TOC to link to our config article, and layout as Overview, Providers, and Implement custom
- Add two new TOC entries for future work, Generic Host and Options pattern

Fixes #763, contributes to #4567

### Preview

✔️ [Configuration in .NET](https://review.docs.microsoft.com/en-us/dotnet/core/extensions/configuration?branch=pr-en-us-20598)
✔️ [Configuration providers in .NET](https://review.docs.microsoft.com/en-us/dotnet/core/extensions/configuration-providers?branch=pr-en-us-20598)
✔️ [Implement a custom configuration provider in .NET](https://review.docs.microsoft.com/en-us/dotnet/core/extensions/custom-configuration-provider?branch=pr-en-us-20598)

### CODE REVIEW

**Comment by BillWagner** [comment]
closing & reopening for a fresh CI build.

File: docs/core/extensions/snippets/configuration/worker-service-options/WorkerWithOptions.cs [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace WorkerServiceOptions.Example
+{
+    public class WorkerWithOptions : BackgroundService
+    {
+        bool _hasWrittenSettings;
+        readonly ILogger<WorkerWithOptions> _logger;
+        readonly CustomSettings _settings;
```
Comment (BillWagner) [inline docs/core/extensions/snippets/configuration/worker-service-options/WorkerWithOptions.cs:14]: I'd prefer these to be explicitly `private`:

```suggestion
        private bool _hasWrittenSettings;
        private readonly ILogger<WorkerWithOptions> _logger;
        private readonly CustomSettings _settings;
```

File: docs/core/extensions/snippets/configuration/custom-provider/Providers/EntityConfigurationSource.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
+﻿using System;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.Extensions.Configuration;
+
+namespace CustomProvider.Example.Providers
+{
+    public class EntityConfigurationSource : IConfigurationSource
+    {
+        readonly Action<DbContextOptionsBuilder> _optionsAction;
```
Comment (BillWagner) [inline docs/core/extensions/snippets/configuration/custom-provider/Providers/EntityConfigurationSource.cs:9]: I'd prefer this to be explicitly private:

```suggestion
        private readonly Action<DbContextOptionsBuilder> _optionsAction;
```

File: docs/core/extensions/snippets/configuration/custom-provider/Providers/EntityConfigurationProvider.cs [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+using System.Linq;
+using CustomProvider.Example.Models;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.Extensions.Configuration;
+
+namespace CustomProvider.Example.Providers
+{
+    public class EntityConfigurationProvider : ConfigurationProvider
+    {
+        readonly Action<DbContextOptionsBuilder> _optionsAction;
```
Comment (BillWagner) [inline docs/core/extensions/snippets/configuration/custom-provider/Providers/EntityConfigurationProvider.cs:12]:
```suggestion
        private readonly Action<DbContextOptionsBuilder> _optionsAction;
```

File: docs/core/extensions/snippets/configuration/worker-service/OptionsWorker.cs [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Logging;
+
+namespace WorkerService.Example
+{
+    public class Worker : BackgroundService
+    {
+        readonly ILogger<Worker> _logger;
```
Comment (BillWagner) [inline docs/core/extensions/snippets/configuration/worker-service/OptionsWorker.cs:11]:
```suggestion
        private readonly ILogger<Worker> _logger;
```

File: docs/core/extensions/configuration-providers.md [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+
+## File configuration provider
+
+The <xref:Microsoft.Extensions.Configuration.FileConfigurationProvider> is the base class for loading configuration from the file system. The following configuration providers derive from `FileConfigurationProvider`:
+
+- [INI configuration provider](#ini-configuration-provider)
+- [JSON configuration provider](#json-configuration-provider)
+- [XML configuration provider](#xml-configuration-provider)
+
+### INI configuration provider
```
Comment (BillWagner) [inline docs/core/extensions/configuration-providers.md:29]: I might consider putting this one last, as it's less of a standard format.

File: docs/core/extensions/configuration-providers.md [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+  <section>
+    <key attribute="value" />
+  </section>
+</configuration>
+```
+
+The previous configuration file loads the following keys with `value`:
+
+- key:attribute
+- section:key:attribute
```
Comment (BillWagner) [inline docs/core/extensions/configuration-providers.md:139]: Should these be code-fenced so they aren't localized?

**Comment by Rick-Anderson** [comment]
in the asp.net core repo, we like to format code without horizontal scroll bars. Test in the widest possible window without TOC - this stimulates tablets

image

File: docs/core/extensions/custom-configuration-provider.md [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+description: Learn how to implement a custom configuration provider in .NET applications.
+author: IEvangelist
+ms.author: dapine
+ms.date: 09/15/2020
+ms.topic: how-to
+---
+
+# Implement a custom configuration provider in .NET
+
+There are many [configuration providers](configuration-providers.md) available for common configuration sources such as INI, JSON, and XML files. You may need to implement a custom configuration provider when one of the available providers doesn't suit your application needs. In this article, you'll learn how to implement a custom configuration provider that relies on a database as its configuration source.
```
Comment (BillWagner) [inline docs/core/extensions/custom-configuration-provider.md:12]: Here again, I'd reorder the configuration sources:

```suggestion
There are many [configuration providers](configuration-providers.md) available for common configuration sources such as JSON, XML, and INI files. You may need to implement a custom configuration provider when one of the available providers doesn't suit your application needs. In this article, you'll learn how to implement a custom configuration provider that relies on a database as its configuration source.
```

File: docs/core/extensions/custom-configuration-provider.md [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+The provider has the following characteristics:
+
+- The EF in-memory database is used for demonstration purposes. To use a database that requires a connection string, implement a secondary `ConfigurationBuilder` to supply the connection string from another configuration provider.
+- The provider reads a database table into configuration at startup. The provider doesn't query the database on a per-key basis.
+- Reload-on-change isn't implemented, so updating the database after the app starts has no effect on the app's configuration.
+
+Define a `Settings` record type entity for storing configuration values in the database.
+
+*Models/Settings.cs*:
```
Comment (BillWagner) [inline docs/core/extensions/custom-configuration-provider.md:26]: I'm not a fan of this format for code samples. It looks a bit odd to me. How about something like this:

```suggestion
Define a `Settings` record type entity for storing configuration values in the database. For example, you could add a *Settings.cs* file in your *Models* folder:
```

**Review by BillWagner — APPROVED** [review]
There's a lot of really good work here @IEvangelist 

I had a number of optional suggestions for you to consider, and then it's ready to :shipit:

**Review by tdykstra — APPROVED** [review]
These articles are well done and the code examples are clear and helpful. I especially like the way the providers doc is structured and code examples provided and explained.  The overview could use some more explanations of terms that are likely to be unfamiliar to people new to the subject.

File: docs/core/extensions/configuration.md [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+<!-- TODO:
+
+- Overview of Microsoft.Extensions article
+- Configuration / Config providers
+- Dependency injection article
+- Generic Host article
+- Logging article
+- Options pattern article
+
+-->
```
[the remark this answers is not in the document]
Comment (IEvangelist) [reply docs/core/extensions/configuration.md:19]: It was an internal note, I can remove it.

File: docs/core/extensions/configuration-providers.md [hunk 10]
```diff
@@ -0,0 +1,316 @@
...
+
+```dotnetcli
+set CustomPrefix__SecretKey="Secret key with CustomPrefix_ environment"
+set CustomPrefix_TransientFaultHandlingOptions__Enabled=true
+set CustomPrefix_TransientFaultHandlingOptions__AutoRetryDelay=00:00:21
+
+dotnet run
+```
+
+The default configuration loads environment variables and command-line arguments prefixed with `DOTNET_` and `ASPNETCORE_`. The `DOTNET_` and `ASPNETCORE_` prefixes are used by .NET for host and app configuration, but not for user configuration.
```
[the remark this answers is not in the document]
Comment (IEvangelist) [reply docs/core/extensions/configuration-providers.md:239]: Yes, it is something that is done regardless - and if there are any env vars that are prefixed they are used... even without that `ConfigureWebHostDefaults` call.

File: docs/core/extensions/configuration-providers.md [hunk 11]
```diff
@@ -0,0 +1,316 @@
...
+<!-- For more information on host and app configuration, see .NET Generic Host. -->
+
+On [Azure App Service](https://azure.microsoft.com/services/app-service), select **New application setting** on the **Settings > Configuration** page. Azure App Service application settings are:
+
+- Encrypted at rest and transmitted over an encrypted channel.
+- Exposed as environment variables.
+
+<!-- For more information, see [Azure Apps: Override app configuration using the Azure Portal](xref:host-and-deploy/azure-apps/index#override-app-configuration-using-the-azure-portal). -->
```
[the remark this answers is not in the document]
Comment (IEvangelist) [reply docs/core/extensions/configuration-providers.md:247]: That is why I commented it out, I'll delete it.

File: docs/core/extensions/configuration-providers.md [hunk 12]
```diff
@@ -0,0 +1,316 @@
...
+| Environment variable key | Converted configuration key | Provider configuration entry                                                    |
+|--------------------------|-----------------------------|---------------------------------------------------------------------------------|
+| `CUSTOMCONNSTR_{KEY}`   | `ConnectionStrings:{KEY}`   | Configuration entry not created.                                                |
+| `MYSQLCONNSTR_{KEY}`     | `ConnectionStrings:{KEY}`   | Key: `ConnectionStrings:{KEY}_ProviderName`:<br>Value: `MySql.Data.MySqlClient` |
+| `SQLAZURECONNSTR_{KEY}`  | `ConnectionStrings:{KEY}`   | Key: `ConnectionStrings:{KEY}_ProviderName`:<br>Value: `System.Data.SqlClient`  |
+| `SQLCONNSTR_{KEY}`       | `ConnectionStrings:{KEY}`   | Key: `ConnectionStrings:{KEY}_ProviderName`:<br>Value: `System.Data.SqlClient`  |
+
+### Environment variables set in launchSettings.json
+
+Environment variables set in *launchSettings.json* override those set in the system environment.
```
[the remark this answers is not in the document]
Comment (IEvangelist) [reply docs/core/extensions/configuration-providers.md:274]: No, *launchSettings.json* are also part of the property system in VS proper as well. I'm not sure if they're available to worker services or not, but I'll verify.

File: docs/core/extensions/configuration-providers.md [hunk 13]
```diff
@@ -0,0 +1,311 @@
...
+<!-- For more information on host and app configuration, see .NET Generic Host. -->
+
+On [Azure App Service](https://azure.microsoft.com/services/app-service), select **New application setting** on the **Settings > Configuration** page. Azure App Service application settings are:
+
+- Encrypted at rest and transmitted over an encrypted channel.
+- Exposed as environment variables.
+
+### Connection string prefixes
+
+The Configuration API has special processing rules for four connection string environment variables. These connection strings are involved in configuring Azure connection strings for the app environment. Environment variables with the prefixes shown in the table are loaded into the app with the [default configuration](#default) or when no prefix is supplied to `AddEnvironmentVariables`.
```
Comment (gewarren) [inline docs/core/extensions/configuration-providers.md:211]: This line introduced a build warning @IEvangelist .

Comment (IEvangelist) [reply docs/core/extensions/configuration-providers.md:211]: I'll fix that, thanks

**Comment by Rick-Anderson** [comment]
@IEvangelist  can we get this merged? I need to refer to it.

**Comment by IEvangelist** [comment]
@Rick-Anderson - This was merged already - it is published once daily... will be live tomorrow morning.

**Pull request merged**